adding ordination as option

trsouz opened the pull request:
Hello ... First thank you for Developing the "monq"!

Now I will try to explain the problem that caused me to develop this "pull-request".

I have a task in the "worker" which is very simple and runs very fast, but I have to climb the worker to have many processors simultaneously.

When climb these "worker" and they try to get tasks at the same time, some can not get any job because "findAndModify" returns only the first one asked. The result is that even if 30 "workers" ready to work only one of them is working and they alternate each other, but only one at a time works.

See example:

``` javascript
'use strict';

var monq = require('monq');

var client = monq('mongodb://admin@mongo:27017/myqueue/?', {
  server: {
    auto_reconnect: true,
    // define poolSize 10 to simulate the concurrency
    poolSize: 10
  }
});

client.db.collection('jobs', function (err, collection) {
  if (err) {
    console.error(err);
  } else {
    // Loop 10 times to simulate the concurrency
    for (var i = 0; i < 10; i++) {
      collection.findAndModify({
        status: 'queued',
        queue: 'mysqueue',
        delay: {$lte: new Date()}
      }, [['priority', 'desc'], ['_id', 'asc']], {
        $set: {
          status: 'dequeued',
          dequeued: new Date()
        }
      }, {new: true}, function (err, doc, other) {
        if (err) {
          console.err(err);
        } else if(doc){
          console.log(doc);
        }else{
          console.log(other);
        }
      });
    }
  }
});
```

If you withdraw the order ([['priority', 'desc'], ['_id', 'asc']]) of "query", the mongo then proceeds to answer to everyone.
So this "pull-request" makes it possible to specify the sort of way you prefer.

(Sorry, my english is not good!)

allancarvalho commented:
+1

scttnlsn commented:
What do you mean by "climb the worker"?  I'm not sure I understand how changing/removing the sorting resolves the issue.  Could you explain a bit more?

trsouz commented:
Sure!!! I mean: "scale the worker"

When 10 workers make the search at the same time, the mongo only responds to the first who made the search. Changing/removing the order in the query, mongo responds to all workers.
Not the best way, but I not found another way to do it.
But this change will not impact the normal operation of the "monq" because it is optional specify the order or not.

Feel free to ask more!
Thanks!

scttnlsn commented:
Thanks, that makes sense.  Is this a known limitation of Mongo?  I thought the whole point of `findAndModify` was to allow concurrent (and atomic) updates like this.  Is it documented somewhere that including a `sort` locks out other processes?  Seems strange to me.

trsouz commented:
I agree with you, that's very strange for me too. The "findAndModify" seems to be the best option to make it so.

I have expressed myself badly, the mongo does not block requests, it responds to all but only returns the document to the first. Note the [method "dequeue" on line "94" when the mongo does not return document](https://github.com/scttnlsn/monq/blob/master/lib/queue.js#L94).

I created a test to be clear the problem I'm pointing (https://github.com/trsouz/monq/tree/concurrency)
You need docker and docker-compose installed to run. This test raises 5 containers to process
First run:

    $ docker-compose pull
    $ docker-compose run --rm node npm install

Then run the test:

    $ ./test_concurrency.sh

This will run four test stages:

  1. Run 10k tasks **with sort** and without delay
  2. Run 10k tasks **with sort** and delay 1-100ms / per task
  3. Run 10k tasks **without sort** and without delay
  4. Run 10k tasks **without sort** and delay 1-100ms / per task

The result is something like *(hh/mm/ss)*:
  1. 00:05:29h
  2. 00:09:36h
  3. 00:00:38h
  4. 00:01:24h

What I observed is that only a "worker" is working at a time when the "sort" is specified.
The result of this problem is that even if you have 100 workers, only one will be working at a time, what is waste of resources.

#### Comments:
If you run the test several times will notice a difference between the results because the delay is randomly generated within the range of 1-100ms for  the test2 and test4.

But even with this difference, is always remarkable the difference between the tasks with "sort" and without the "sort".

Tests can be configured in the file "docker-compose-concurrency.yml" and the tasks are stored in different collections to be analyzed after the tests.

Thank you for your patience and your time!

scttnlsn commented:
So is there any way to maintain the priority functionality here?

markstos commented:
I would translate the problem like this. The MongoDB `findAndModify` feature is like the SQL `SELECT...FOR UPDATE` feature. 

PostgreSQL 9.5 added "SELECT..FOR UPDATE...SKIP LOCKED", which sounds like what is desired from MongoDB here, but here missing.  More discussion about this feature in use for job queues is discussed here:

http://stackoverflow.com/a/13005551/254318

(PostgreSQL supports LISTEN and NOTIFY, so you don't have to poll the database, the database can notify each interested worker when there is work to do).

scttnlsn commented:
@markstos That sounds like exactly what's missing here.

I've been out of touch with MongoDB for a while- is anyone aware of a solution to this?

danielkrainas commented:
There was something I read online that if you added an index to the target sort field (for `priority` in this case), the problem doesn't persist. I can experiment tonight to see if that works.

scttnlsn commented:
@danielkrainas That'd be awesome if you figure this out.  Keep me posted.  Thanks!

trsouz commented:
Hi guys, sorry my disappearance.
Last week I updated the "monq" to the latest version and removed my amendment on the "ordination". Throughout the week I accompanied my "workers" and did not report any error in relation to the competition between them.

I'm able to process about 17 to 25 activities per second with 10 "workers".

But in fact I have not done all the tests I'd like to ensure that the problem was actually solved.

All of you are using the latest version?

scttnlsn commented:
@trsouz What version were you using previously?  Not sure what would have changed.

trsouz commented:
@scttnlsn I was using the v0.3.2 version. I suspect that the update of mongoskin and mongodb have influence.

scttnlsn commented:
Please re-open if this is ever an issue again.  Further discussion will take place in https://github.com/scttnlsn/monq/issues/55
